Include/cpython/object.h

```diff
@@ -320,6 +320,17 @@ PyAPI_FUNC(int)
 _PyObject_GenericSetAttrWithDict(PyObject *, PyObject *,
                                  PyObject *, PyObject *);
 
+/* Replacements of PyObject_GetItem() which don't raise KeyError.
+
+   Return 1 and set *result != NULL if a key is found.
+   Return 0 and set *result == NULL if a key is not found;
+   a KeyError is silenced.
+   Return -1 and set *result == NULL if an error other than KeyError
+   is raised.
+*/
+PyAPI_FUNC(int) _PyMapping_LookupItem(PyObject *, PyObject *, PyObject **);
+PyAPI_FUNC(int) _PyMapping_LookupItemString(PyObject *, const char *, PyObject **);
```

Comment:
Can you please add it to the internal C API instead, since it's a private API?

Reply:
I added it next to `_PyObject_LookupAttr`, isn't it the correct place? What is better?

Reply:
`_PyObject_LookupAttr` is in the awkward legacy "in-between state" that currently causes some controversy, where it is marked private by initial underscore (though I guess this wasn't clearly documented as a marker of private API until relatively recently?), but it is part of `Python.h` and publicly visible, so easily usable by extensions, meaning we have to worry whether it is implicitly public anyway and we can't touch it without causing churn.

By "add to internal C API" @vstinner means to put it in `Include/internal/pycore_object.h` (where it will only be usable by extensions setting `Py_BUILD_CORE` and explicitly including that header, in which case they clearly opt in to using private API).

But personally I think this should be explicitly public API, as I mentioned on the issue.

Reply:
Thank you for clarification @carljm. I always make this mistake: `Include/cpython/` looks like something deeply internal to me.

`_PyObject_LookupAttr` never was purposed to use outside of Python. I though that undersoring the name and explicitly excluding it from the limited C API was enough red herring.

Now it will be in the limited C API (#106521) as well as this function.